### PR TITLE
PLT-604 Add PriceIndex

### PR DESCRIPTION
**Pull request by jodastephen**
Adds a `PriceIndex` and implementations that match the design of Fx, Ibor and Overnight. Selected indices are included for major currencies.

### CODE REVIEW

File: modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+  /**
+   * The index name, such as 'GB-HICP'.
+   */
+  @PropertyDefinition(validate = "notEmpty", overrideGet = true)
+  private final String name;
+  /**
+   * The region of the index.
+   */
+  @PropertyDefinition(validate = "notNull", overrideGet = true)
+  private final Country region;
```
Comment (julianhowarth) [inline modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java]: Doesn't this suggest either Country or this field is misnamed?

File: modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+ * A price index implementation based on an immutable set of rules.
+ * <p>
+ * A standard immutable implementation of {@link PriceIndex}.
+ * <p>
+ * In most cases, applications should refer to indices by name, using {@link PriceIndex#of(String)}.
+ * The named index will typically be resolved to an instance of this class.
+ * As such, it is recommended to use the {@code PriceIndex} interface in application
+ * code rather than directly referring to this class.
+ */
+@BeanDefinition(cacheHashCode = true)
```
Comment (julianhowarth) [inline modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java]: Why are we caching the hashCode for this particular object? Or are we moving to a position where we think all immutable objects should set this?

File: modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java (hunk 1 above)
Comment (jodastephen) [inline modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java]: Country is based on the ISO-3166-1 name which includes Europe, and Europe is the only region of note here.

File: modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java (hunk 2 above)
Comment (jodastephen) [inline modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java]: Beans that are hashmap keys should have cached hash codes. Over time, we'll see if that is most beans. Right now, we know this one will be a hash key.

**Pull request merged**